repl/functions.go

```diff
@@ -734,108 +734,155 @@ func MakeUserFunction(name string, ufun GlispUserFunction) *SexpFunction {
-	"cons":       ConsFunction,
-	"first":      FirstFunction,
-	"second":     SecondFunction,
-	"rest":       RestFunction,
-	"car":        FirstFunction,
-	"cdr":        RestFunction,
-	"type":       TypeQueryFunction,
-	"list?":      TypeQueryFunction,
-	"null?":      TypeQueryFunction,
-	"array?":     TypeQueryFunction,
-	"hash?":      TypeQueryFunction,
-	"number?":    TypeQueryFunction,
-	"int?":       TypeQueryFunction,
-	"float?":     TypeQueryFunction,
-	"char?":      TypeQueryFunction,
-	"symbol?":    TypeQueryFunction,
-	"string?":    TypeQueryFunction,
-	"zero?":      TypeQueryFunction,
-	"empty?":     TypeQueryFunction,
-	"println":    PrintFunction,
-	"print":      PrintFunction,
-	"printf":     PrintFunction,
-	"not":        NotFunction,
-	"apply":      ApplyFunction,
-	"map":        MapFunction,
-	"make-array": MakeArrayFunction,
-	"aget":       ArrayAccessFunction,
-	"aset!":      ArrayAccessFunction,
-	"sget":       SgetFunction,
-	"hget":       GenericAccessFunction, // handles arrays or hashes
-	"hset!":      HashAccessFunction,
-	"hdel!":      HashAccessFunction,
-	"keys":       HashAccessFunction,
-	"hpair":      GenericHpairFunction,
-	"slice":      SliceFunction,
-	"len":        LenFunction,
-	"append":     AppendFunction,
-	"concat":     ConcatFunction,
-	"array":      ConstructorFunction,
-	"list":       ConstructorFunction,
-	"hash":       ConstructorFunction,
-	"msgmap":     ConstructorFunction,
-	"raw":        ConstructorFunction,
-	"raw2str":    RawToStringFunction,
-	"symnum":     SymnumFunction,
-	"source":     SourceFileFunction,
-	//"source":    SimpleSourceFunction,
-	"str2sym":   Str2SymFunction,
-	"sym2str":   Sym2StrFunction,
-	"gensym":    GensymFunction,
-	"str":       StringifyFunction,
-	"->":        ThreadMapFunction,
-	"json":      JsonFunction,
-	"unjson":    JsonFunction,
-	"msgpack":   JsonFunction,
-	"unmsgpack": JsonFunction,
-	"togo":      ToGoFunction,
-	"dump":      GoonDumpFunction,
-	"slurpf":    SlurpfileFunction,
-	"writef":    WriteToFileFunction,
-	"owritef":   WriteToFileFunction,
-	"system":    SystemFunction,
-	"flatten":   FlattenToWordsFunction,
-	"nsplit":    SplitStringOnNewlinesFunction,
-	"split":     SplitStringFunction,
-	"methodls":  GoMethodListFunction,
-	"_method":   CallGoMethodFunction,
-	"fieldls":   GoFieldListFunction,
-	"chomp":     StringUtilFunction,
-	"trim":      StringUtilFunction,
-	"exit":      ExitFunction,
-	"stop":      StopFunction,
-	"_closdump": DumpClosureEnvFunction,
-	"_call":     CallZMethodOnRecordFunction,
-	"gob":       GobEncodeFunction,
-	"=":         AssignmentFunction,
-	"joinsym":   JoinSymFunction,
-	"quotelist": QuoteListFunction,
-	"rmsym":     RemoveSymFunction,
-	"GOOS":      GOOSFunction,
+// mergeFuncMap returns the union of the two given maps
+func mergeFuncMap(funcs ...map[string]GlispUserFunction) map[string]GlispUserFunction {
+	n := make(map[string]GlispUserFunction)
+
+	for _, f := range funcs {
+		for k, v := range f {
+			n[k] = v
```

Comment:
I thought about that but came up with a few issues.

1. This doesn't allow for overriding built in functions without manually monkey patching out those functions. I can see this being done for mocking during tests etc.
2. Can't use in function signature of factories due to multiple return values.

2 is purely a convenience thing. Maybe two versions of the function should be available?